core/runtime/src/main/java/io/quarkus/runtime/ApplicationConfig.java

```diff
@@ -0,0 +1,31 @@
+package io.quarkus.runtime;
+
+import io.quarkus.runtime.annotations.ConfigItem;
+import io.quarkus.runtime.annotations.ConfigPhase;
+import io.quarkus.runtime.annotations.ConfigRoot;
+
+@ConfigRoot(phase = ConfigPhase.BUILD_AND_RUN_TIME_FIXED)
+public class ApplicationConfig {
+
+    /**
+     * The name of the application.
+     * If not set, defaults to the name of the project.
+     */
+    @ConfigItem(defaultValue = "${project.name:__UNSET__}")
+    public String name;
```

Comment:
If we allow it to be unset, it should just be optional.

Reply:
Can a property have a default value and also be `Optional`?

Reply:
The reason I went this was is because of https://github.com/quarkusio/quarkus/blob/master/core/deployment/src/main/java/io/quarkus/deployment/configuration/RunTimeConfigurationGenerator.java#L738. 
With the current code of `ApplicationConfig` in master the line above is what (I believe) causes the value to not be used at runtime.

Reply:
> Can a property have a default value and also be `Optional`?

Yes, it can.  But the default value should be meaningful.  It looks like you might be trying to hack around something the wrong way here.

> The reason I went with this is, is because of https://github.com/quarkusio/quarkus/blob/master/core/deployment/src/main/java/io/quarkus/deployment/configuration/RunTimeConfigurationGenerator.java#L738.
> With the current code of `ApplicationConfig` in master the line above is what (I believe) causes the value to not be used at runtime.

This logic essentially means "if this property has no default value, and no other sub-trees of this value have default values, don't generate pointless default value methods which would always return null for every name".  This logic only controls what values end up as defaults at run time based on what's in the parse tree.  If a property has no default value, then no default value needs to be returned for it.

This logic does *not* generate run-time-readable config values for values which were specified in build time config, which I suspect is what you're trying to do here.  Trying to use this logic for that purpose will yield incorrect results.

Reply:
Well I wasn't trying to use something in any particular way. I was lead to that piece of code by debugging and comparing what made this type of configuration different from others that did behave the way we expected. 
But let me go back and take another look.

Reply:
> This logic does _not_ generate run-time-readable config values for values which were specified in build time config, which I suspect is what you're trying to do here. 

And yes, that is exactly the end result I am looking for :)

Reply:
Thanks for the invaluable input! I think this iteration is better. WDYT?